errors.go

```diff
@@ -0,0 +1,527 @@
+package httpio
+
+import (
+	stderrors "errors"
+	"fmt"
+	"strings"
+
+	"github.com/go-playground/errors/v5"
+)
+
+type msgType int
+
+const (
+	badRequest          msgType = iota // http code 400
+	unauthorized                       // http code 401
+	forbidden                          // http code 403
+	notFound                           // http code 404
+	conflict                           // http code 409
+	internalServerError                // http code 500
+	serviceUnavailable                 // http code 503
+)
+
+func init() {
+	errors.RegisterErrorFormatFn(errorFormatFn)
+}
+
+// errorFormatFn implements a custom format function for the errors package
+// to properly unwrap error chains contained inside a ClientMessage.
+func errorFormatFn(chain errors.Chain) string {
+	var errMsg strings.Builder
+
+	for i, link := range chain {
+		if i == 0 {
+			if chain, ok := stderrors.Unwrap(link.Err).(errors.Chain); ok {
+				errMsg.WriteString(errorFormatFn(chain))
+				errMsg.WriteString("\n")
+			}
+		}
+		if i > 0 {
+			errMsg.WriteString("\n")
+		}
+		errMsg.WriteString(link.Error())
+	}
+
+	return errMsg.String()
+}
+
+// Message returns the message from the wrapped ClientMessage or an empty string
+func Message(err error) string {
+	cerr := &ClientMessage{}
+	if errors.As(err, &cerr) {
+		return cerr.Message()
+	}
+
+	return ""
+}
+
+// Messages returns a slice of messages from the ClientMessage's contained within the chain of errors
+func Messages(err error) []string {
+	cerr := &ClientMessage{}
+	if errors.As(err, &cerr) {
+		subMsgs := Messages(cerr.Unwrap())
+
+		msgs := make([]string, 0, len(subMsgs)+1)
+		msgs = append(msgs, cerr.Message())
+		msgs = append(msgs, subMsgs...)
+
+		return msgs
+	}
+
+	return nil
+}
+
+// CauseIsError returns true if the Cause of this error is an error vs a ClientMessage with nil error
+func CauseIsError(err error) bool {
+	cerr := &ClientMessage{}
+	if errors.As(err, &cerr) {
+		return CauseIsError(cerr.Unwrap())
+	}
+
+	return err != nil
+}
+
+// ClientMessage is a custom message type that can be used to return client messages
+type ClientMessage struct {
+	msgType       msgType
+	clientMessage string
+	error         error
+}
+
+// Message returns the client message
+func (c *ClientMessage) Message() string {
+	return c.clientMessage
+}
+
+// Error returns the error message
+func (c *ClientMessage) Error() string {
+	return c.clientMessage
+}
```

Comment:
Maybe the comments here could be improved? The comments make it sound like the two methods will return different things, with the second potentially involving c.error... But they are both simply returning the client message.

Reply:
Done